### ISSUE TITLE
[Feature] GKE Kubernetes Cluster Manangement

### ISSUE THREAD

**Issue by gambol99** [appvia/appvia-hub#163]
Looking into the possibility of cluster creation and management of GKE cluster _(initially)_, but in general cloud provider implementations.   

* Do we create the project?
  * should the gcp project be created or we assume it exists
* Do we set best practice on the organization via policies?
* Do the we enable the APIs on the users behalf or assume they've done it
* If we've created the account, how should billing be handled
  * do we show a list of billing accounts at the org level
  * do we request the requirements for a new billing account
* What are the options for creating a cluster?
  * region, machine, size, node
  * cluster type zonal, regional, etc
  * shared vpc?
  * project name
* In regard to machine type, defaults types?
  * do we allow them to choose the types?
  * do we allow them use customizes instances?
  * are options should as GPUs and types exposed?
  * preemptible VMs?
* How are addons managed? a checkbox?
* How would you manage users
  * are they local?
  * are they organizational?
* how do they login post creation?
* How would they manage node groups?
  * back to machine types?
  * taints, labels howto?
* Security options
  * checkbox for psp?
  * options to workload identity? maps GSA -> KSA

**Describe the solution you'd like**
Users, either globally or per space should be about to create and manage clusters within a cloud provider. We don't want to reinvent the console ui of a provider, so the direction should be somewhat prescribed in terms of configurations _(i.e. upfront choice which provide best practice should be made on the user's behalf)_

**What value would this add?**
* Provides a single portal / jump off point for users to manage the clusters within the cloud.

**Comment by jon-shanks** [comment]
So i think for the hub, a lot of elements would no doubt be out of scope, primarily because the outcome would be to have a kubernetes cluster for consumption for the hub resources. 

So assumptions we can probably make are:
- They already have GCP, (a pre-req of wanting to use GKE provider)
- They have an org setup / project
- They have suitable privileges to be able to create a GKE
- They have permissions to create service accounts that grant the relevant permissions or a pre-req could be that it is a service account

We probably need to understand more about what we would want this cluster to look like and then what its intended use would be?

- Would we expect this cluster to be used for applications only?
- Is it a flat cluster with ops tooling + applications
- How would someone interact with the cluster after it is built?

It would be good to draw up a bit of a proposal doc around the above and some healthy assumptions / defaults and then work from there.

**Comment by gambol99** [comment]
https://docs.google.com/document/d/1suFNK663qzs2BtMSkQWTB1oC7qed0GkcWQZl6YDWHDc/edit

**Comment by danielwhatmuff** [comment]
Noting some thoughts on what I'd expect from this feature and some possible solutions based on past discovery work as discussed earlier.

When creating a cluster I would expect to be able to input a name, select from a list of options for cluster level components (e.g. prometheus, loki and grafana) and some best practice security options (e.g. RBAC/PSPs) to be installed/enabled upon its creation.

Once the cluster has been created I would expect to be able to use it as part of the kubernetes integration to create namespaces for deploying my apps as well as provisioning logging and metric dashboards using the grafana and loki integrations.

Given the complexity involved in kube cluster lifecycle management it may be wise initially to hand off to GKE anything other than just create/delete by providing an external link through to the GKE console and/or providing steps to install and manage clusters using the gcloud CLI.

In order to avoid having to track and reconcile the list and status of GKE clusters which have been created using the hub and which could have since been terminated or updated outside of the hub (e.g. via GCP console) I would be inclined to consider performing the cluster creation in an async manner and then making requests to the GKE API to list clusters in the GCP project and get their status at the point where a user navigates to a cluster view within the hub. The API https://cloud.google.com/resource-manager/reference/rest/v1/operations/get could potentially also be used in order to track the async cluster creation.

**Comment by danielwhatmuff** [comment]
@jits I realise the above may not fit into the current resource model. What are your thoughts on that? We could store just the name and description of a cluster upon creation and then use GETs on the GCP APIs to determine the state of it post creation.

**Comment by jits** [comment]
@danielwhatmuff – 

(I've mentioned this a couple of times in planning sessions + [somewhat in comments in the doc](https://docs.google.com/a/appvia.io/document/d/1suFNK663qzs2BtMSkQWTB1oC7qed0GkcWQZl6YDWHDc/edit?disco=AAAADGhEGVc) linked above).

For a first version / MVP deliverable I don't think we should use the current resources + provider integrations + agents model, as I think that will be a bit too heavyweight an approach, **and I don't believe we need to store the state of the clusters in the hub, and delve in to cluster _management_ via the hub**.

Instead I think we should see it as a different capability for leaner _one-off bootstrapping_  of cluster(s) via some interface in the hub, only by hub admins, with the sole output of setting up the correct integrations in the hub – i.e. Kubernetes + Grafana + Loki integrations configured in the hub – and the cluster would have some initial best practices enabled (which we would inform the hub admin as part of the creation process).

I'll explain my rationale below…

**My understanding is there are two key goals here:**

1. Figure out and iterate on what we can do, and mean by, "best practices" clusters, that fit in with the capabilities provided by the hub (e.g. Grafana and Loki dashboards) (i.e. Rohith's doc above)
2. An MVP implementation in the hub to serve two key use cases:
  a) Make it easier to set up an instance of the hub with running GKE cluster(s) with _some_ best practices.
  b) Have a functional basis for gathering user feedback (i.e. an MVP implementation that's initially useful, but can help trigger further user input through actual usage).

**Assumptions we could make right now:**

- We don't need to support project/space specific clusters yet – we can stick to the current way integrations work (available to all projects/spaces).
- We can defer the updating and lifecycle management of clusters, and see if there really is a need for this in the hub, or if this can be done outside of the hub once the cluster is created.
- We should avoid building full on cluster management in the hub!

**Proposal**

Based on this, the simplest possible MVP implementation I can currently think of, for us to iterate on after, is:
- Have a new hub admin section in the hub called 'Bootstrap' (exact phrasing TBD), or reuse the existing 'Integrations' section…
  - … with a button to "Create a new GKE cluster with best practices".
- Clicking on this button takes you to a form / one page wizard…
  - … here you _may_ be able to toggle on/off certain features of the cluster (TBD)
  - … and you're given a detailed breakdown of what this will do.
- Submitting this will trigger a new background job…
  - … which can use a Ruby script within hub codebase to perform all the necessary actions to set this cluster up + dependencies (e.g. Grafana, Loki).
- If _intermediate_ state and/or user intervention is required (e.g. to set up the Grafana auth), then we can store this in a separate db table, or in Redis (i.e. not using the current resources model), and report back to the hub admin the progress of the job…
  - … this allows us to build the underlying job in a more asynchronous way (i.e. break it down into a workflow / sequence of jobs)
- Once finished…
  - … the script would return back all the necessary config data required by the hub to instantiate / setup the necessary integrations
  - … we can display the status of this job (assuming we're storing the intermediate state for it) on the 'Bootstrap' page for future reference
- We wouldn't need a list of clusters because we already have the list of Kubernetes integrations on the Integrations page + the list of jobs that were carried out (successfully or not; assuming we store intermediate state for these jobs)

As you can see, there's a lot of implementation details that would need to be figured out, but the key goal is to not delve into full on cluster management via the hub – i.e. once clusters are created, the hub doesn't do anything else about them.

**Comment by jits** [comment]
@danielwhatmuff – I realise reading back that you're not explicitly saying we should use the resources etc. models currently in the hub… so the latter half of my comment is probably more relevant :)

**Comment by danielwhatmuff** [comment]
It sounds like we're pretty much on the same page based on your comment above apart from two things.

1. I would say an admin user _would_ expect to be able to see the cluster/clusters that have been created from the hub somewhere within the hub and not for it to be abstracted away fully just into a configured kubernetes integration. I think that would be too much of a hub specific logical leap for them to know whats going on. Also would allow admins to easily go to GKE to resize/debug/view logs if there are issues.

2. I'm not sure about it being a "one-off" thing as clusters by their nature are usually created/updated/destroyed/created as and when required. Sometimes deleted out of hours to save hosting costs or test building platforms from scratch each day to have confidence in disaster recovery. We have also already heard of use cases such as having a cluster per team/environment/application and moving away from multi-tenancy.

**Comment by jits** [comment]
For (1) - maybe; would be good to validate that hypothesis somehow before committing to it. But if it's just a case of maintaining a list of the clusters that were created / failed, then that's very doable (and in line with the jobs state thing I described above). Anything more...

... like (2) sounds like a big ask for the hub to take on. I don't dispute the _need_ and the use cases you described, but it sounds like a lot to commit to for an initial MVP implementation. My preference would be to start off with the one-off approach and take it from there, based on actual usage and customer feedback. No doubt those needs will come up, and when they do we may be able to cater for them in a more creative / cost effective / useful way (e.g. use an existing cluster management product).

**Comment by danielwhatmuff** [comment]
1 - Sure understood. Just to note also, my thoughts on it aren't just based on personal preference, it's also based on research of other products.

2 - If it's the simpler option to keep it as one-off then it makes sense to do that initially and iterate from there based on feedback.

**Comment by jits** [comment]
@jon-shanks mentioned something we need to think about: how do things like GKE + GCR coupling affect the MVP implementation? I.e. since GCR is tied to a GCP project, I guess the assumption is that any GKE clusters would need to be created within the same project for private images access within the cluster. Since Docker Repos are space specific, maybe all it would take is for the cluster creation tool to output the config for a GCR Docker Repo integration?

There may be other cases like this, that could affect design decisions right now.

**Comment by gambol99** [comment]
So in essence _(forgoing something that pops up in some whiteboard session)_ .. a script _(bash would be quickest, as it mean not messing about with ruby sdk and api - though will mean adding the kubectl and gcloud binaries to the hub container)_; with build and destroy commands - a bunch of command line / environment vars options controlling features ... which

a) builds cluster
b) potentially deploying some default grafana and loki manifests (i'm assuming oauth2 stuff is passed on the command line and is env var via secret for the deployment) from somewhere / url / chart 
c) and outputs some kubecfg options .. potentials endpoints for grafana; though that would means having a zone, scraping the k8s service for the load balancer name and adding as a cname into some know domain - tls will be an issue. Or just passing back the load balancer hostname for now

**Comment by danielwhatmuff** [comment]
on b) the Grafana OAuth stuff is configured using a `.ini` file read by the service. I was using the stable helm package and passing that config using a values file: https://github.com/helm/charts/blob/d2860446f7759edcb0ed1bc3e07e9c4aaff8226e/stable/grafana/values.yaml#L340

e.g.
```
[auth.github]
enabled = true
allow_sign_up = true
client_id = ...
client_secret = ...
scopes = user,read:org
auth_url = https://github.com/login/oauth/authorize
token_url = https://github.com/login/oauth/access_token
api_url = https://api.github.com/user
allowed_organizations = a-github-org
```

**Comment by danielwhatmuff** [comment]
loki and prometheus can be done using the default charts:
```
$ helm install loki/loki-stack -n loki
$ helm install stable/prometheus
```

**Comment by gambol99** [comment]
The changes to the gem, custom errors, wait on job, docker etc have been done ..

**Comment by gambol99** [comment]
@jits .. I've added the various features requests, a json schema which can be called by `instance.schema.to_json` or just hands back the hash, added custom errors and pulling the defaults from the schema. Adding the dns related bit for granfana and exposed the github creds in the options. This schema also has tags to provide a layout of options

**Comment by jits** [comment]
As part of this work, I'll be fixing the known issue of Grafana and Loki integrations not being tied directly to a Kubernetes integration (which only impacts once you have more than one cluster).

**Comment by jits** [comment]
TODO:

- [ ] Update deployments with [new set of background workers](https://github.com/appvia/appvia-hub#background-workers)
  - [ ] For AWS demo hub
  - [ ] For the customer demos

/cc @davet1985 (when you're back)
